### PR TITLE
[DO NOT MERGE] [CPF-4671] combining methods for replacing characters for json outputter 

### PR DESCRIPTION
**Pull request by olhabn**
https://nanopay.atlassian.net/browse/CPF-4671

### CODE REVIEW

File: src/foam/lib/json/Outputter.java [hunk 1]
```diff
@@ -100,56 +108,65 @@ public void outputString(String s) {
...
+          if(c >= ' ') {
+            if(i == s.length() - 1)
+              appendSubString(writer, s, startIndex, s.length());
+          }
+          else {
+            appendSubString(writer, s, startIndex, i);
+            startIndex = i + 1;
+            char right = (char) (c & 0x0F);
+            char left = (char) ((c & 0xF0) >> 4);
+            right += '0';
```
Comment (kgrgreer) [inline src/foam/lib/json/Outputter.java]: += on strings is very expensive because it builds a temporary StringBuilder each time. Can you refactor to not use String arithmetic?

File: src/foam/lib/json/Outputter.java [hunk 2]
```diff
@@ -34,6 +35,13 @@ protected SimpleDateFormat initialValue() {
       return df;
     }
   };
+  protected static Map<Character, String> charactersThatNeedToBeReplaced = new HashMap() {{
```
Comment (kgrgreer) [inline src/foam/lib/json/Outputter.java]: Faster to just use a switch statement.

File: src/foam/lib/json/Outputter.java [hunk 3]
```diff
@@ -100,56 +108,65 @@ public void outputString(String s) {
...
+    try {
+      char c;
+      int startIndex = 0;
+
+      for ( int i = 0; i < s.length(); i++ ) {
+        c = s.charAt(i);
+
+        String replaceWithCharacter = charactersThatNeedToBeReplaced.get(c);
+        if(replaceWithCharacter != null) {
+          appendSubString(writer, s, startIndex, i);
```
Comment (kgrgreer) [inline src/foam/lib/json/Outputter.java]: Why is this needed? Couldn't you just copy each character as you go rather than remembering ranges and creating substrings?

Comment (olhabn) [reply src/foam/lib/json/Outputter.java]: Fixed

File: src/foam/lib/json/Outputter.java (hunk 2 above)
Comment (olhabn) [reply src/foam/lib/json/Outputter.java]: Fixed

File: src/foam/lib/json/Outputter.java [hunk 4]
```diff
@@ -111,45 +111,9 @@ public void outputString(String s) {
...
-      // Add previous string segment
-      escapedString += s.substring(lastStart, i);
-      // Add escape sequence
-      escapedString += new String(escape);
-      lastStart = i + 1;
-    }
-    if ( lastStart != s.length() ) escapedString +=
-      s.substring(lastStart, s.length());
-    return escapedString;
+    StringBuilder sb = new StringBuilder();
```
Comment (kgrgreer) [inline src/foam/lib/json/Outputter.java]: Don't bother with this, since we're just going to replace it with the new code anyway.

File: src/foam/lib/json/Util.java [hunk 5]
```diff
@@ -11,8 +11,46 @@
...
+            if(c >= ' ')
+              dst.append(c);
+            else {
+              char right = (char) (c & 0x0F);
+              char left = (char) ((c & 0xF0) >> 4);
+              right += '0';
+              if ( right > '9' ) right += 'A' - '9' - 1;
+              left += '0';
+              if ( left > '9' ) left += 'A' - '9' - 1;
+              dst.append('\\');
```
Comment (kgrgreer) [inline src/foam/lib/json/Util.java]: Next four lines could be replaced with dst.append("\\u00");

File: src/foam/lib/json/Util.java [hunk 6]
```diff
@@ -11,8 +11,46 @@
...
+              dst.append('u');
+              dst.append('0');
+              dst.append('0');
+              dst.append(left);
+              dst.append(right);
+            }
+            break;
```
Comment (kgrgreer) [inline src/foam/lib/json/Util.java]: Do you need this break?

File: src/foam/lib/json/Util.java [hunk 7]
```diff
@@ -11,8 +11,46 @@
...
+            dst.append("\\n");
+            break;
+          case '\\':
+            dst.append("\\\\");
+            break;
+          case '\"':
+            dst.append("\\\"");
+            break;
+          default:
+            if(c >= ' ')
```
Comment (kgrgreer) [inline src/foam/lib/json/Util.java]: Follow FOAM style guide for spacing.
